src/assisted_installer_controller/assisted_installer_controller.go

```diff
@@ -70,22 +75,41 @@ func NewController(log *logrus.Logger, cfg ControllerConfig, ops ops.Ops, ic inv
 	}
 }
 
-func (c *controller) WaitAndUpdateNodesStatus() {
+func (status *ControllerStatus) Error() {
+	atomic.AddUint32(&status.errCounter, 1)
+}
+
+func (status *ControllerStatus) HasError() bool {
+	return atomic.LoadUint32(&status.errCounter) > 0
+}
+
+func (c *controller) WaitAndUpdateNodesStatus(status *ControllerStatus) {
 	c.log.Infof("Waiting till all nodes will join and update status to assisted installer")
-	ignoreStatuses := []string{models.HostStatusDisabled,
-		models.HostStatusError, models.HostStatusInstalled}
+	ignoreStatuses := []string{models.HostStatusDisabled, models.HostStatusInstalled}
+	var hostsInError int
 	for {
 		time.Sleep(GeneralWaitInterval)
 		ctx := utils.GenerateRequestContext()
 		log := utils.RequestIDLogger(ctx, c.log)
+
 		assistedInstallerNodesMap, err := c.ic.GetHosts(ctx, log, ignoreStatuses)
 		if err != nil {
 			log.WithError(err).Error("Failed to get node map from inventory")
 			continue
 		}
+		errNodesMap := c.ic.FilterHostsByStatus(assistedInstallerNodesMap, models.HostStatusError)
+		hostsInError = len(errNodesMap)
+
+		//if all hosts are in error, mark the failure and finish
+		if hostsInError > 0 && hostsInError == len(assistedInstallerNodesMap) {
```

Comment:
i think it is not the right behavior, what if only 1 worker is in error?

Reply:
so actually we should ask here "if all master hosts are in error"? basically it was meant to catch the situation of timeout where all hosts are moved to "error" state

Reply:
yeah, got it